src/main/java/frc/robot/subsystems/Intake.kt

```diff
@@ -63,7 +72,46 @@ object Intake : Subsystem {
     val hasBranchCoral: Boolean
         get() = canRange.isDetected.value
 
+    val supplyCurrent: Current
+        get() = leader.supplyCurrent.value
+
+    val torqueCurrent: Current
+        get() = leader.torqueCurrent.value
+
+    /**
+     * Return true if the intake motor is experiencing current draw greater than the given
+     * threshold.
+     */
+    fun hasCoralByCurrent(): Boolean {
+        // return true if current is spiked and coral is detected by CANRange
+        return (leader.torqueCurrent.value >= CORAL_CURRENT_THRESHOLD) && hasBranchCoral
+    }
+
+    fun hasAlgaeByCurrent(): Boolean {
+        val thresholdMet: Boolean = leader.torqueCurrent.value >= ALGAE_CURRENT_THRESHOLD
+        if (thresholdMet) {
+            if (algaeSpikeDetected) {
+                if (algaeSpikeResolved) {
+                    // if threshold is met, and initial current spike from running motor is already
+                    // accounted for,
+                    // we have the algae so return true and reset the spike detection booleans
+                    algaeSpikeDetected = false
+                    algaeSpikeResolved = false
+                    return true
+                }
+            } else {
+                // threshold met and initial spike not detected yet, so mark it as such
+                algaeSpikeDetected = true
+            }
+        } else if (algaeSpikeDetected) {
+            // if spike was detected but current is back down, is now resolved
+            algaeSpikeResolved = true
+        }
+        return false
```

Comment:
Definitely want to simplify this in the future to just base off current and maybe velocity. Trivial to implement a low-pass filter that will ignore the first spike, but trip on the constant spike later on.